imports/plugins/included/shipping-rates/server/no-meteor/util/filterShippingMethods.js

```diff
@@ -0,0 +1,46 @@
+import { attributeDenyCheck } from "./attributeDenyCheck";
+import { locationAllowCheck } from "./locationAllowCheck";
+import { locationDenyCheck } from "./locationDenyCheck";
+
+
+/**
+ * @summary Filter shipping methods based on per method restrictions
+ * @param {Object} context - an object containing the per-request state
+ * @param {Object} methods - all available shipping methods for a shop
+ * @param {Object} hydratedOrder - hydrated order for current order
+ * @returns {Object|null} available shipping methods after filtering
+ */
+export default async function filterShippingMethods(context, methods, hydratedOrder) {
+  const flatRateFulfillmentRestrictionsCollection = context.collections.FlatRateFulfillmentRestrictions;
+
+  const allValidShippingMethods = methods.reduce(async (validShippingMethods, method) => {
+    const awaitedValidShippingMethods = await validShippingMethods;
+
+    // If method is not enabled, it is not valid
+    if (!method.enabled) {
+      return awaitedValidShippingMethods;
+    }
+
+    // Find all restrictions for this shipping method
+    const methodRestrictions = await flatRateFulfillmentRestrictionsCollection.find({ methodIds: method._id }).toArray();
+
+    // Check method against location allow check
+    const allowedMethodBasedOnShippingLocationsAllowList = await locationAllowCheck(methodRestrictions, method, hydratedOrder);
+
+    // Check method against location deny check
+    const allowedMethodsBasedOnShippingLocationsDenyList = await locationDenyCheck(methodRestrictions, method, hydratedOrder);
```

Comment:
Also, it will be more efficient if you exit as soon as you know it's not allowed. So I would add an early exit line after each of these, and then remove the IF that's currently around `awaitedValidShippingMethods.push`.

```js
const methodIsAllowedBasedOnShippingLocationsAllowList = await locationAllowCheck(methodRestrictions, method, hydratedOrder);
if (!methodIsAllowedBasedOnShippingLocationsAllowList) return awaitedValidShippingMethods;
// and so on
```